### PR TITLE
Added a score for assessing vulnerability discovery and security testing

### LINKED ISSUE #224: Assessing number of vulnerabilities together with security testing
A number of vulnerabilities and their severities can't tell much about security in an open-source project. However, it may make sense to assess info about vulnerabilities together with info about security testing implemented in the project. It may be implemented in a score that has the following logic:

|   |  Bad testing | Good testing |
|---|---|---|
| Few vulnerabilities | Low score | Max score  |
| Many vulnerabilities | Min score  | High score  |

The logic is quite simple. It may be improved, for example, by taking into account severity of vulnerabilities.

### LINKED ISSUE #262: AbstractScore.equals() should consider descriptions
https://github.com/SAP/fosstars-rating-core/blob/b13419621942538ab06887963583ff011be10522/src/main/java/com/sap/sgs/phosphor/fosstars/model/score/AbstractScore.java#L134

The `AbstractScore.equals()` method doesn't use `that.description`. That should be fixed.

### LINKED ISSUE #264: Better vulnerability definition
There are a few issues in the `Vulnerability` class:

- Some constructors are redundant. The `Builder` should be used instead.
- The builder should use CVSS v3 if available.
- The `Vulnerability` class should store a date when the issue was published.

### LINKED ISSUE #265: Not applicable values should be known
https://github.com/SAP/fosstars-rating-core/blob/d6d287c77a5084c73c7fb8020845402525d8bfb1/src/main/java/com/sap/sgs/phosphor/fosstars/model/value/NotApplicableValue.java#L35

The method should return true.

### LINKED ISSUE #266: Confidence.make() should consider all possible values
The `Confidence.make()` method currently counts only unknown values. However, score values have their own confidence and weight. Furthermore, a value may implement the `Confidence` interface. Therefore. the `make()` method needs to be updated to consider all these cases.

### PR DESCRIPTION
**Pull request by artem-smotrakov**
Here is a list of main updates:

- Implemented `VulnerabilityDiscoveryAndSecurityTestingScore`.
- Added the new score to the open-source security score.
- Improved `Vulnerability` class and its builder. This fixes #264
- Fixed #262 
- Fixed #265 
- Improved calculation of confidence. This fixes #266

This closes #224

### CODE REVIEW

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/json/UnpatchedVulnerabilitiesStorage.java [hunk 1]
```diff
@@ -170,42 +168,14 @@ private static UnpatchedVulnerabilitiesStorage check(UnpatchedVulnerabilitiesSto
...
         "https://github.com/odata4j/odata4j",
-        new Vulnerability(
-            "https://nvd.nist.gov/vuln/detail/CVE-2014-0171",
-            "XML external entity (XXE) vulnerability",
-            CVSS.v2(5.0),
-            Collections.emptyList(),
-            Resolution.UNPATCHED,
-            UNKNOWN_INTRODUCED_DATE,
-            UNKNOWN_FIXED_DATE));
-    storage.add(
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/json/UnpatchedVulnerabilitiesStorage.java:183]: Are we not adding any more test CVEs?

File: src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java [hunk 2]
```diff
@@ -66,28 +69,60 @@ static double adjust(double value) {
...
+   *   <li>number of unknown values</li>
+   *   <li>score values with their confidences and weights</li>
+   *   <li>values with confidences</li>
+   * </ul>
    *
    * @param values The features values.
    * @return The confidence level.
    */
-  static double make(Value... values) {
+  static double make(List<Value> values) {
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java:83]: We use `SetValue` almost every place. Does it make sense to have a common representation everywhere?

File: src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java [hunk 3]
```diff
@@ -66,28 +69,60 @@ static double adjust(double value) {
...
+        weightSum += 1.0;
+      } else if (value instanceof ScoreValue) {
+        ScoreValue scoreValue = (ScoreValue) value;
+        weightSum += scoreValue.weight();
+        weightedConfidenceSum += scoreValue.weight() * scoreValue.confidence();
+      } else if (value instanceof Confidence) {
+        weightedConfidenceSum += ((Confidence) value).confidence();
+        weightSum += 1.0;
+      } else {
+
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java:106]: Please remove this extra line.

File: src/main/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScore.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  /**
+   * A description of the score.
+   */
+  private static final String DESCRIPTION
+      = "The scores checks how security testing is done\n"
+      + "and how many vulnerabilities were recently discovered.\n"
+      + "If testing is good, and there are no recent vulnerabilities,\n"
+      + "then the score value is max.\n"
+      + "If there are vulnerabilities, then the score value is high.\n"
+      + "If testing is bad, and there are not recent vulnerabilities,\n"
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScore.java:53]: there are no recent vulnerabilities

File: src/test/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScoreTest.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import com.sap.sgs.phosphor.fosstars.model.other.Utils;
+import com.sap.sgs.phosphor.fosstars.model.value.ScoreValue;
+import com.sap.sgs.phosphor.fosstars.model.value.Vulnerabilities;
+import java.util.Date;
+import java.util.Set;
+import org.junit.Test;
+
+public class VulnerabilityDiscoveryAndSecurityTestingScoreTest {
+
+  private static ProjectSecurityTestingScore SECURITY_TESTING_SCORE
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScoreTest.java:16]: Is it ok to use the same object while testing in different methods?

File: src/test/java/com/sap/sgs/phosphor/fosstars/model/value/VulnerabilitiesValueTest.java [hunk 6]
```diff
@@ -14,8 +15,10 @@
   public void serialization() throws IOException {
     ObjectMapper mapper = new ObjectMapper();
     VulnerabilitiesValue vulnerabilityValue =
-        new VulnerabilitiesValue(new Vulnerabilities(new Vulnerability("https://bugtracker/1"),
-            new Vulnerability("https://bugtracker/2")));
+        new VulnerabilitiesValue(
+            new Vulnerabilities(
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/sgs/phosphor/fosstars/model/value/VulnerabilitiesValueTest.java:19]: If `VulnerabilitiesValue` itself contains list of `Vulnerabilities`, then do we need to do a wrapper here?

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/json/UnpatchedVulnerabilitiesStorage.java (hunk 1 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/json/UnpatchedVulnerabilitiesStorage.java:183]: This code is not actually for adding data for tests. It's just a convenient method for updating the JSON file with unpatched issues.

File: src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java (hunk 2 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java:83]: Do you mean `Set<Value>` or something else? Once we need `make(Set<Value>)` method, we can definitely add it. Currently, the `make()` method is used with a `List` returned by `ScoreValue.usedValues()`.

File: src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java (hunk 3 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/model/Confidence.java:106]: Actually I intentionally added a line here. See also above. For some reason, an extra line before a comment in the beginning of a block looks nicer to me. I know, it's quite subjective. I can remove it if it bothers.

File: src/main/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScore.java (hunk 4 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScore.java:53]: Sure.

File: src/test/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScoreTest.java (hunk 5 above)
Comment (artem-smotrakov) [reply src/test/java/com/sap/sgs/phosphor/fosstars/model/score/oss/VulnerabilityDiscoveryAndSecurityTestingScoreTest.java:16]: As long as the object is immutable, or at least has an expected state before before a test case runs, I think yes, it's okay.

File: src/test/java/com/sap/sgs/phosphor/fosstars/model/value/VulnerabilitiesValueTest.java (hunk 6 above)
Comment (artem-smotrakov) [reply src/test/java/com/sap/sgs/phosphor/fosstars/model/value/VulnerabilitiesValueTest.java:19]: `VulnerabilitiesValue` contains an instance of `Vulnerabilities` which is a collection of vulnerabilities. In fact, it doesn't use a list, it uses a set. I know that `Vulnerabilities` may look redundant but it also provide multiple convenient methods to work with a collection of vulnerabilities. Of course, we can add a constructor `VulnerabilitiesValue(List ... )` if necessary.

**Comment by artem-smotrakov** [comment]
Thanks for the review @sourabhsparkala !

**Pull request merged**